File: leetcode/42. Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int l = 0, r = n - 1, maxLeft = 0, maxRight = 0;
        int ans = 0;
        while (l <= r) {
            if (height[l] <= height[r]) {
                if (height[l] >= maxLeft) {
                    maxLeft = height[l];
                } else {
                    ans = ans + maxLeft - height[l];
                }
                l++;
            } else {
                if (height[r] >= maxRight) {
                    maxRight = height[r];
                } else {
                    ans = ans + maxRight - height[r];
                }
                r--;
            }
        }
        return ans;
    }
};

File: leetcode/42. Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        l, r = 0, n - 1
        maxLeft, maxRight = 0, 0
        ans = 0
        
        while l <= r:
            if height[l] <= height[r]:
                if height[l] >= maxLeft:
                    maxLeft = height[l]
                else:
                    ans += maxLeft - height[l]
                l += 1
            else:
                if height[r] >= maxRight:
                    maxRight = height[r]
                else:
                    ans += maxRight - height[r]
                r -= 1
        
        return ans